code-monkey/src/agents/agent.py

```diff
@@ -1,78 +1,148 @@
 import os
+from typing import Set
+from langgraph.prebuilt import create_react_agent
+from langchain_core.messages import HumanMessage, SystemMessage
+from langchain_core.language_models.chat_models import BaseChatModel
+
 from tools.utils import (
     ask_user,
     show_diff,
 )
-from constants import get_src_dir, get_artifacts_dir
-from models import Model, get_model_service
+from constants import get_root_dir, get_artifacts_dir
+from models import parse_msn
 from .base_agent import BaseAgent
 from instrumentation import current_span, instrument
+from langgraph.checkpoint.aiosqlite import AsyncSqliteSaver
 
 class Agent(BaseAgent):
-    model: Model
+    model: BaseChatModel
+
+    config = {
+        "configurable": {"thread_id": "abc123"},
+    }
 
+    @instrument("Agent.__init__", ["msn"])
     def __init__(self, msn: str | None):
-        # TODO(toshok) these two steps should be collapsed and hidden back in the
-        # models/ directory (and replaced with something like this here:
-        #   self.model = construct_model(msn, self)
-        Service = get_model_service(msn)
-        self.model = Service(self, msn)
+        [ ChatModel, model, extra_flags ] = parse_msn(msn)
+
+        # a checkpointer + the thread_id below gives the model a way to save its
+        # state so we don't have to accumulate the messages ourselves.
+        memory = AsyncSqliteSaver.from_conn_string(":memory:")
+        self.model = create_react_agent(
+            ChatModel(
+                model_name=model,
+                extra_flags=extra_flags,
+            ),
+            self.tools,
+            checkpointer=memory,
+        )
+        self.initialize()
 
     # Custom Agent initialization goes here, when necessary.
     def initialize(self):
         pass
 
-    def run_prompt(self, prompt: str):
+    @instrument("Agent.run_prompt", ["prompt"])
+    async def run_prompt(self, prompt: str):
         import logging
         logger = logging.getLogger(__name__)
         logger.debug(f"Running prompt: {prompt}")
-        result = self.model.run_prompt(prompt)
-        logger.debug(f"Prompt result: {result}")
-        return result
+
+        system = SystemMessage(content=self.get_system_prompt())
+        msg = HumanMessage(content=self.prepare_prompt(prompt))
+
+        modified_files: Set[str] = set()
+
+        async for event in self.model.astream_events({
+            "messages": [system, msg],
+        }, self.config, version="v2"):
+            kind = event["event"]
+            if kind == "on_chat_model_start":
```

Comment:
Is this ai-generated code with all these `if`s instead of an event map or dict?

Reply:
no that's manual.  I wasn't sure I needed/wanted, and what behavior we'd want so it felt like the wrong thing to pull it all out initially.